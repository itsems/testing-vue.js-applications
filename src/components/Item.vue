<template>
  <div>
    <a :href="item.url">{{ item.title }}</a>
    <span class="host">
      {{ item.url | host }}
    </span>
    <span class="score">{{ item.score }}</span>
    {{ item.author }}
    {{ item.time | timeAgo }} ago
    <span v-if="item.type !== 'job'" class="by">
      type is not 'job'
    </span>
      
      
  </div>
</template>

<script>
export default {
  props: ['item']
}
</script>