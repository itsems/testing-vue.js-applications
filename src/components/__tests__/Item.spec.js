import { shallowMount } from '@vue/test-utils'
import Item from '../Item.vue'


describe('Item', () => {
  test('renders the hostname', () => {
    const item = {
      url: 'https://some-url.com/with-paths'
    }
    const wrapper = shallowMount(Item, {
      propsData: { item }
    });
    expect(wrapper.text()).toContain('some-url.com')
  })
  test('render item.title', () => {
    const item = {
      title: 'my title',
    }
    const wrapper = shallowMount(Item, {
      propsData: { item }
    });
    expect(wrapper.find('a').text()).toBe(item.title)
  })
  test('renders a link to item.url with item.title as text', () => {
    const item = {
      url: 'https://medium.com/itsems-frontend',
      title: 'My Medium'
    }

    const wrapper = shallowMount(Item, {
      propsData: { item }
    })
    const a = wrapper.find('a')
    expect(a.text()).toBe(item.title)
    expect(a.attributes().href).toBe(item.url)

  })

  // chapter-3 練習 1
  test('render item.score & item.author', () => {
    const item = {
      score: 50,
      author: 'emma',
    }
    const wrapper = shallowMount(Item, {
      propsData: { item }
    })
    expect(wrapper.text()).toContain(item.score)
    expect(wrapper.text()).toContain(item.author)
  })

  test('renders the time since the last post', () =>{
    const dateNow = jest.spyOn(Date, 'now')
    const dateNowTime = new Date('2018')
    dateNow.mockImplementation(() => dateNowTime)

    const item = {
      time: (dateNowTime / 1000) - 600 //10分鐘前的UNIX
    }
    const wrapper = shallowMount(Item, {
      propsData: {
        item
      }
    })
    dateNow.mockRestore()
    expect(wrapper.text()).toContain('10 minutes ago')
  })

})