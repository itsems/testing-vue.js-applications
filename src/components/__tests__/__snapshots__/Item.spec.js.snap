// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item renders correctly 1`] = `
<div>
  <a
    href="https://www.google.com"
  >
    hello
  </a>
   
  <span
    class="host"
  >
    
    google.com
  
  </span>
   
  <span
    class="score"
  >
    100
  </span>
  
  emma
  10 minutes ago
  
  <span
    class="by"
  >
    
    type is not 'job'
  
  </span>
</div>
`;

exports[`Item renders correctly as job 1`] = `
<div>
  <a
    href="https://www.google.com"
  >
    hello
  </a>
   
  <span
    class="host"
  >
    
    google.com
  
  </span>
   
  <span
    class="score"
  >
    100
  </span>
  
  emma
  10 minutes ago
  
  <!---->
</div>
`;
